lib/logstash/inputs/tcp/decoder_impl.rb

```diff
@@ -11,24 +11,32 @@ def initialize(codec, tcp)
     @first_read = true
   end
 
-  def decode(channel_addr, data)
+  def decode(ctx, data)
+    channel_addr = ctx.channel().remoteAddress()
     bytes = Java::byte[data.readableBytes].new
     data.getBytes(0, bytes)
     data.release
     tbuf = String.from_java_bytes bytes, "ASCII-8BIT"
     if @first_read
       tbuf = init_first_read(channel_addr, tbuf)
     end
-    @tcp.decode_buffer(@ip_address, @address, @port, @codec,
-                       @proxy_address, @proxy_port, tbuf, nil)
+    if @tcp.ssl_enable && @tcp.ssl_verify
+      session = ctx.channel().pipeline().get("ssl-handler").engine().getSession()
```

Comment:
this is a pretty long chain, and it is unclear why we stop here to bind a local variable `session`, which we use only once to continue the chain.

 - are all methods in the chain documented to never return `nil`?
 - should we extract this complexity to a helper method that has appropriate `rescue` clauses?

Reply:
Hey @yaauie! I stopped there mostly as a means to follow the flow here: https://github.com/logstash-plugins/logstash-input-beats/blob/5dd54594f65d32aad87d1dfd7b04d0c801216676/lib/logstash/inputs/beats/message_listener.rb#L125-L155

I'm happy to add in the rescue that's covered here: https://github.com/logstash-plugins/logstash-input-beats/blob/5dd54594f65d32aad87d1dfd7b04d0c801216676/lib/logstash/inputs/beats/message_listener.rb#L132-L139

Reply:
> are all methods in the chain documented to never return nil?

It seems that way, aside from the `ChannelPipeline#get` we invoke, though since we explicitly name our channelhandler [here](https://github.com/logstash-plugins/logstash-input-tcp/pull/159/files#diff-ed2ceebf1bfbd6e8dad3c53f7b0c53e0R111), we should be safe in that respect.

>should we extract this complexity to a helper method that has appropriate rescue clauses?

I think thats a great idea, though I sort of think it might be worth tucking in somewhere like the `logstash/base/inputs` package, since we'd be repeating the same logic in this plugin, the beats input plugin, and potentially other plugins where netty is used and TLS metadata extraction is possible. We would need to be careful about formatting though, as different input plugins seem to serialize TLS metadata differently. (it gets cut to 3 fields in the beats plugin and one field here).

Also, as an update on my last message, looking more closely, I'm not sure that rescue clause I cite would really be adding much in terms of safety, but would be more for debug logging insight.